refact_data_pipeline/finetune/supported_models.py

```diff
@@ -151,5 +151,43 @@
         },
         "train_model_modifiers": [],
         "force_enable_checkpointing": True
+    },
+
+    "codellama/7b": {
+        "lora_target_modules_mapping": {
+            "qkv": ["attn.q_attn", "attn.c_attn"],
+            "out": ["attn.c_proj"],
+            "backproj": ["attn.c_proj"],
+            "mlp": ["mlp.c_fc", "mlp.c_proj"],
+        },
+        "freeze_exceptions_mapping": {
+            "wte": "wte",
+            "lm_head": "lm_head",
+            "lora": "lora"
+        },
+        "tokenizer": {
+            "eot_idx": 0,
```

Comment:
I think there is a problem since those indexes are for `bigcode` encoding, and you are using a llama encoding

They also use slightly different encoding format (with `bos` token) and maybe the `FIM` format is also different

Reply:
Sorry, just saw this. I understand what you are saying, but I'm a little unclear on how to find the correct encoding. This is the [Code Llama tokenizer](https://huggingface.co/docs/transformers/main/model_doc/code_llama#transformers.CodeLlamaTokenizer). Should I create a new `code_llama.json` In the refact_encoding folder with these tokens? Where did the `bigcode_largemodel.json` file come from? Where did the vocab part of the file come from? When I google the entire file for Code Llama the best result is [this](https://huggingface.co/MarioPenguin/beto_amazon/commit/83bd1c1ba9422bc57e90b92f93b4bd112afea8be.diff?file=tokenizer.json) which is for BERT.

Reply:
We are using encodings which are downloaded from HF directly. Since you're using https://huggingface.co/TheBloke/CodeLlama-7B-fp16, try to study the [tokenizer file](https://huggingface.co/TheBloke/CodeLlama-7B-fp16/raw/main/tokenizer.json), [code llama FIM format](https://github.com/ggerganov/llama.cpp/issues/2818). Compare them with some official sources. 
As an `escape` token you can choose any "empty" token, treat it like some kind of "control" symbol

Also, don't forget about `BOS` token, as I remember llama models are using it as a starting label. If it's true - make sure our datapipeline prepares data correctly

Reply:
Yes, that makes sense. I will look into it.

Alternatively, you are sure that lines 135-143 in encoding.py, are not a valid method of applying the llama tokenization?

```
        elif name in ['llama']:
            from sentencepiece import SentencePieceProcessor
            filename = Path(__file__).resolve().parent / f"{name}.tokenizer.model"
            self._sentencepiece_tokenizer = SentencePieceProcessor(str(filename))
            self.n_vocab = self._sentencepiece_tokenizer.vocab_size()
            self.bos_id: int = self._sentencepiece_tokenizer.bos_id()
            self.DIAMOND = self._sentencepiece_tokenizer.unk_id()
            self.EOT = self._sentencepiece_tokenizer.eos_id()
            self.LF = 13
```

Reply:
It should be valid. 
However, we don't use this code unless you're using an old `codify` model
See https://github.com/smallcloudai/refact/blob/main/refact_data_pipeline/finetune/model_handling.py#L127 to get the idea


I must mention, that the logic behind constructing all these objects is weird, that's why we are refactoring it right now (https://github.com/smallcloudai/refact/pull/182)